// fits_canvas/lib/display.js
// Generated by CoffeeScript 1.3.3
(function() {
  var Display;

  Display = (function() {

    function Display(container, width, height) {
      this.container = container;
      this.width = width;
      this.height = height;
      this.canvasContainer = document.getElementById(this.container);
      this.canvas = document.createElement('canvas');
      this.canvas.width = this.width;
      this.canvas.height = this.height;
      this.canvasContainer.appendChild(this.canvas);
      this.context = this.canvas.getContext('2d');
      this.buildImageBuffers();
    }

    Display.prototype.buildImageBuffers = function() {
      this.displayData = this.context.createImageData(this.width, this.height);
      this.displayBuffer = new ArrayBuffer(this.displayData.data.length);
      this.displayView8 = new Uint8ClampedArray(this.displayBuffer);
      this.displayView32 = new Uint32Array(this.displayBuffer);
    };

    Display.prototype.draw = function() {
      this.displayData.data.set(this.displayView8);
      this.context.putImageData(this.displayData, 0, 0);
    };

    return Display;

  })();

  if (typeof module !== "undefined" && module !== null) {
    module.exports = Display;
  }

}).call(this);
